### DISCUSSION TITLE
Whether the format and pattern keywords can exist simultaneously under the string type

### DISCUSSION THREAD

**Discussion by suncpp** [json-schema-org/community#262] [Q&A] [answered]
code:
`{
    "type": "string",
    "pattern":"regx",
    "format":"regx",
}`
When format and pattern are both used as assertions, can they exist at the same time? Which has the highest priority?

**Comment by handrews** [comment, marked as answer]
@suncpp all JSON Schema keywords can be used together, and they all have the same priority.  Some of these combinations make sense, and some do not (e.g. `minimum` is superfluous in `{"type": "string", "minimum": 2}` because any number will fail the `type` assertion), but none are forbidden.

`pattern` and `format` are often used together because depending on the draft and configuration you are using, `format` may or may not behave as an assertion.  And even if it does, that keyword is pretty unreliable as it's not implemented consistently.

So people often use `format` as a semantic annotation (its default behavior in 2020-12) and use `pattern` to do the actual validation.  e.g. `{"type": "string", "format": "email", "pattern": "some-ridiculous-email-validiating-regexp"}`.  In that case, `format` serves more as machine-readable documentation.

But if you do configure things so that `format` has assertion behavior, then both it and `pattern` have equal priority in the sense that they both must pass for the schema to pass.  If either fails, then the schema fails.

**Reply by suncpp** [reply]
Thank you for your reply.
